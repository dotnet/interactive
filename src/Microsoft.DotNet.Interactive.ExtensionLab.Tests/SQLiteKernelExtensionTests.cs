// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;

using Assent;
using Microsoft.DotNet.Interactive.CSharp;
using Microsoft.DotNet.Interactive.Events;
using Microsoft.DotNet.Interactive.Formatting;
using Microsoft.DotNet.Interactive.Tests.Utility;

using Xunit;

namespace Microsoft.DotNet.Interactive.ExtensionLab.Tests
{
    public class SQLiteKernelExtensionTests : IDisposable
    {
        private readonly Configuration _configuration;

        public SQLiteKernelExtensionTests()
        {
            _configuration = new Configuration()
                .SetInteractive(Debugger.IsAttached)
                .UsingExtension("json");
        }

        [Fact]
        public async Task can_generate_tabular_json_from_database_table_result()
        {
            using var kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
                new KeyValueStoreKernel()
            };

            var extension = new SQLiteKernelExtension();

            await extension.OnLoadAsync(kernel);

            using var _ = SQLiteConnectionTests.CreateInMemorySQLiteDb(out var connectionString);

            await kernel.SubmitCodeAsync(
                $"#!connect --kernel-name mydb sqlite \"{connectionString}\"");

            var result = await kernel.SubmitCodeAsync(@"
#!sql-mydb
SELECT * FROM fruit
");

            var events = result.KernelEvents.ToSubscribedList();

            var formattedData = events.OfType<DisplayedValueProduced>().Single()
                .FormattedValues.Single(fm => fm.MimeType == HtmlFormatter.MimeType)
                .Value;

           
            this.Assent(formattedData, _configuration);
        }

        [Fact]
        public async Task can_handle_duplicate_columns_in_query_results()
        {
            using var kernel = new CompositeKernel
            {
                new CSharpKernel().UseNugetDirective(),
                new KeyValueStoreKernel()
            };

            var extension = new SQLiteKernelExtension();

            await extension.OnLoadAsync(kernel);

            using var _ = SQLiteConnectionTests.CreateInMemorySQLiteDb(out var connectionString);

            await kernel.SubmitCodeAsync(
                $"#!connect --kernel-name mydb sqlite \"{connectionString}\"");

            var result = await kernel.SubmitCodeAsync(@"
#!sql-mydb
SELECT 1 AS Apples, 2 AS Bananas, 3 AS Apples, 4 AS BANANAS, 5 AS Apples, 6 AS BaNaNaS
");

            var events = result.KernelEvents.ToSubscribedList();

            var formattedData = events
                                .OfType<DisplayedValueProduced>()
                                .Single()
                                .FormattedValues
                                .Single(fm => fm.MimeType == HtmlFormatter.MimeType)
                                .Value;

            this.Assent(formattedData, _configuration);
        }

        public void Dispose()
        {
            Formatter.ResetToDefault();
            DataExplorer.ResetToDefault();
        }
    }
}