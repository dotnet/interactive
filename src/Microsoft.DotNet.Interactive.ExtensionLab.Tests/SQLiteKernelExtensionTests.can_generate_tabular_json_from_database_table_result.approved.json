<div style="background-color:white;">
<div id="00000000000000000000000000000000" style="height: 100ch ;margin: 2px;">
</div>
<script type="text/javascript">

let renderNteractDataExplorer_00000000000000000000000000000000 = () => {

    (require.config({ 'paths': { 'context': '1.0.252001', 'nteractUri' : 'https://colombod.github.io/dotnet-interactive-cdn/extensionlab/1.0.252001/nteract/nteractapi', 'urlArgs': 'cacheBuster=00000000000000000000000000000000' }}) || require)(['nteractUri'], (nteract) => {

        nteract.createDataExplorer({
            data: {
  "profile": "tabular-data-resource",
  "schema": {
    "primaryKey": [],
    "fields": [
      {
        "name": "id",
        "type": "integer"
      },
      {
        "name": "name",
        "type": "string"
      },
      {
        "name": "color",
        "type": "string"
      },
      {
        "name": "deliciousness",
        "type": "integer"
      }
    ]
  },
  "data": [
    {
      "id": 1,
      "name": "apple",
      "color": "green",
      "deliciousness": 10
    },
    {
      "id": 2,
      "name": "banana",
      "color": "red",
      "deliciousness": 11
    },
    {
      "id": 3,
      "name": "cherry",
      "color": "red",
      "deliciousness": 9000
    }
  ]
},
            container: document.getElementById("00000000000000000000000000000000")
        });
    },
    (error) => {
        console.log(error);
    });
};
// ensure `require` is available globally
if ((typeof(require) !==  typeof(Function)) || (typeof(require.config) !== typeof(Function))) {
    let require_script = document.createElement('script');
    require_script.setAttribute('src', 'https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js');
    require_script.setAttribute('type', 'text/javascript');
    
    
    require_script.onload = function() {
        renderNteractDataExplorer_00000000000000000000000000000000();
    };

    document.getElementsByTagName('head')[0].appendChild(require_script);
}
else {
    renderNteractDataExplorer_00000000000000000000000000000000();
}

 </script>
</div>
